### PR TITLE
Implement RegExp octal support

### PR DESCRIPTION
**Pull request by szledan**
JerryScript-DCO-1.0-Signed-off-by: Dániel Bátyai dbatyai.u-szeged@partner.samsung.com
JerryScript-DCO-1.0-Signed-off-by: Szilard Ledan szledan.u-szeged@partner.samsung.com

### CODE REVIEW

**Comment by LaszloLango** [comment]
@szledan, cloud you add some test for this?

**Comment by dbatyai** [comment]
Hi, I've updated the patch.
I added some test and changed backreferences to try to fall back to octal escapes when they're invalid.
Also note that I've rebased this patch atop #485 for convenience, so please disregard that commit.

File: jerry-core/parser/regexp/re-parser.cpp [hunk 1]
```diff
@@ -71,6 +71,27 @@ re_parse_non_greedy_char (lit_utf8_iterator_t *iter_p) /**< RegExp pattern */
...
+re_parse_octal (lit_utf8_iterator_t *iter) /**< input string iterator */
+{
+  uint32_t number = 0;
+  for (int index = 0;
+       index < 3
+       && !lit_utf8_iterator_is_eos (iter)
+       && lit_char_is_octal_digit (lit_utf8_iterator_peek_next (iter));
+       index++)
+  {
+    number = number * 8 + lit_char_hex_to_int (lit_utf8_iterator_read_next (iter));
```
Comment (zherczeg) [inline jerry-core/parser/regexp/re-parser.cpp]: hex_to_int? I suspect there is no octal, and this decrease code size. Ok, keep it.

File: jerry-core/parser/regexp/re-parser.cpp [hunk 2]
```diff
@@ -744,12 +752,44 @@ re_parse_next_token (re_parser_ctx_t *parser_ctx_p, /**< RegExp parser context *
             {
               out_token_p->type = RE_TOK_BACKREFERENCE;
             }
+            else
+            /* Invalid backreference, fallback to octal */
+            {
+              /* Rewind to start of number. */
+              while (index-- > 0)
+              {
+                lit_utf8_iterator_read_prev (iter_p);
```
Comment (zherczeg) [inline jerry-core/parser/regexp/re-parser.cpp]: There is a function to move back without reading the character

File: jerry-core/parser/regexp/re-parser.cpp [hunk 3]
```diff
@@ -744,12 +752,44 @@ re_parse_next_token (re_parser_ctx_t *parser_ctx_p, /**< RegExp parser context *
...
           {
-            out_token_p->value = ch;
+            if (!lit_char_is_octal_digit (ch))
+            {
+              /* Not octal, keep character value. */
+              out_token_p->value = ch;
+            }
+            else
+            {
+              lit_utf8_iterator_read_prev (iter_p);
```
Comment (zherczeg) [inline jerry-core/parser/regexp/re-parser.cpp]: Ditto

File: tests/jerry/regexp-simple-atom-and-iterations.js [hunk 4]
```diff
@@ -56,3 +56,15 @@ assert (r.exec ("abcdefghijkl") == "abcdefghijkl");
...
+assert (/[\1284]+/.exec ("1\n\n8\n4\n\n2") == "\n\n8\n4\n\n");
+assert (/[\89]12/.exec ("1\9128123") == "912");
+assert (/[\11]/.exec ("1\n\n\t\n\n2") == "\t");
+assert (/[\142][\143][\144]/.exec ("abcde") == "bcd");
+
+assert (/\12+/.exec ("1\n\n\n\n\n2") == "\n\n\n\n\n");
+assert (/\11/.exec ("1\n\n\t\n\n2") == "\t");
+assert (/\142\143\144/.exec ("abcde") == "bcd");
+assert (/\942\143\144/.exec ("a942cde") == "942cd");
+assert (/\14234/.exec ("b34") == "b34");
```
Comment (zherczeg) [inline tests/jerry/regexp-simple-atom-and-iterations.js]: Please add test case which contains capturing brackets and backreferences fall back (or not) to octal parsing.

File: jerry-core/parser/regexp/re-parser.cpp [hunk 5]
```diff
@@ -465,26 +486,13 @@ re_parse_char_class (re_parser_ctx_t *parser_ctx_p, /**< number of classes */
...
-          /* FIXME: octal support */
-        }
-      }
-      /* FIXME: depends on the unicode support
-      else if (!jerry_unicode_identifier (ch))
+               && lit_char_is_octal_digit ((ecma_char_t) ch)
+               && ch != LIT_CHAR_0)
       {
-        JERRY_ERROR_MSG ("RegExp escape pattern error. (Char class)");
+        lit_utf8_iterator_read_prev (iter_p);
```
Comment (zherczeg) [inline jerry-core/parser/regexp/re-parser.cpp]: Ditto

**Comment by dbatyai** [comment]
@zherczeg, patch is updated.

File: tests/jerry/regexp-simple-atom-and-iterations.js [hunk 6]
```diff
@@ -56,3 +56,21 @@ assert (r.exec ("abcdefghijkl") == "abcdefghijkl");
...
+
+print (/(\d+)\2([abc]+)\1\2/.exec("123abc123abc") == "123abc123abc,123,abc");
```
Comment (kkristof) [inline tests/jerry/regexp-simple-atom-and-iterations.js]: I assume these `print` might be `assert`.

Comment (dbatyai) [inline tests/jerry/regexp-simple-atom-and-iterations.js]: Fixed, sorry.

**Comment by zherczeg** [comment]
LGTM

**Comment by sand1k** [comment]
LGTM.

**Pull request merged**